From 102dfb912200ca14e58effe1e4c49db368f17036 Mon Sep 17 00:00:00 2001
From: WofWca <wofwca@protonmail.com>
Date: Tue, 28 Oct 2025 21:36:43 +0400
Subject: [PATCH] fix: allow gibbing after `intermissionQueued`

The argument for this: it's dissatisfying that you can't
gib the corpse after finishing someone with the gauntlet.
It's especially dissatisfying that the last frag,
if it's a shotgun frag is destined to _never_ gib.
The round could end with a bang, but it ends with a whimper.
---
 code/game/g_combat.c | 23 +++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/code/game/g_combat.c b/code/game/g_combat.c
index b8f7448a..beafafb6 100644
--- a/code/game/g_combat.c
+++ b/code/game/g_combat.c
@@ -817,6 +817,29 @@ void G_Damage( gentity_t *targ, gentity_t *inflictor, gentity_t *attacker,
 	// the intermission has already been qualified for, so don't
 	// allow any extra scoring
 	if ( level.intermissionQueued ) {
+
+		// With a special exception for gibbing bodies.
+		// This was introduced in https://github.com/ec-/baseq3a/pull/50.
+		if (targ->die == body_die) {
+			// Note that this also affects the situation where the last frag
+			// is done with the shotgun. Because each pellet calls `G_Damage`
+			// separately. That is, if we don't do this,
+			// the shotgun will never gib if it's the last frag.
+			//
+			// Note that this does not check for friendly fire, quad,
+			// battlesuit, handicap, armor, etc.
+			// We could simply continue normal execution of G_Damage
+			// instead of returning, but let's play it safe
+			// and only do what's necessary.
+
+			targ->health = targ->health - damage;
+			if ( targ->health <= 0 ) {
+				// `body_die` doesn't use any of its arguments (except `targ`),
+				// so it's fine if they're `NULL`.
+				targ->die (targ, inflictor, attacker, damage, mod);
+			}
+		}
+
 		return;
 	}
 #ifdef MISSIONPACK
-- 
2.43.0

